pandas/core/dtypes/cast.py

```diff
@@ -246,6 +246,16 @@ def _disallow_mismatched_datetimelike(value, dtype: DtypeObj):
         raise TypeError(f"Cannot cast {repr(value)} to {dtype}")
 
 
+@overload
+def maybe_downcast_to_dtype(result: np.ndarray, dtype: str | np.dtype) -> np.ndarray:
+    ...
+
+
+@overload
+def maybe_downcast_to_dtype(result: ArrayLike, dtype: str | np.dtype) -> ArrayLike:
```

Comment:
pandas/_typing.py defines `ArrayLike` as `Union["ExtensionArray", np.ndarray]`. If `maybe_downcast_to_dtype` returns an ExtensionArray, when it is given an ExtensionArray, it would be stricter to replace `ArrayLike` with `ExtensionArray`.

@rhshadrach suggestions of using a `ArrayLikeT = TypeVar("ArrayLikeT", bound=ArrayLike)` would be/is a more elegant solution if sub-types of ExtensionArrays are maintained (I don't know whether that is the case).